File: COMP30050-2008/users/BHarty/construct/construct/src/org/construct_infrastructure/client/AbstractClientProxy.java
//
// $Revision: 7871 $: $Date: 2008-02-25 15:06:55 +0000 (Mon, 25 Feb 2008) $ - $Author: gstevenson $
//
//  This file is part of Construct, a context-aware systems platform.
//  Copyright (c) 2006, 2007, 2008 UCD Dublin. All rights reserved.
// 
//  Construct is free software; you can redistribute it and/or modify
//  it under the terms of the GNU Lesser General Public License as
//  published by the Free Software Foundation; either version 2.1 of
//  the License, or (at your option) any later version.
// 
//  Construct is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
//  GNU Lesser General Public License for more details.
// 
//  You should have received a copy of the GNU Lesser General Public
//  License along with Construct; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
//  USA
//
//  Further information about Construct can be obtained from
//  http://www.construct-infrastructure.org
package org.construct_infrastructure.client;


import java.io.IOException;
import java.net.Socket;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeoutException;
import java.util.logging.Level;
import java.util.logging.Logger;

import org.construct_infrastructure.io.Message;
import org.construct_infrastructure.io.MessageReader;
import org.construct_infrastructure.io.ServiceComponentDescriptor;

/**
 * Client-side proxy object that can talk to a Construct discovery service, find services and
 * interact with them.
 * 
 * @author Steve Neely (steve.neely@ucd.ie)
 * @author Graeme Stevenson (graeme.stevenson@ucd.ie)
 */
public abstract class AbstractClientProxy {

   /**
    * The port on which the Construct bonjour proxy is running.
    */
   private static final int LOCAL_CONSTRUCT_BONJOUR_PROXY_PORT = 3826;

   /**
    * The length of the &lt;/serviceComponentDescriptor&gt; tag.
    */
   private static final int SERVICE_COMPONENT_DESCRIPTOR_END_TAG_LENGTH = 29;

   /**
    * The logger for this class.
    */
   private final Logger my_logger;

   /**
    * Creates a new instance of the proxy and starts a bonjour listener.
    */
   public AbstractClientProxy() {
      // Set up the logger
      my_logger = Logger.getLogger(getClass().getName());
      getLogger().setLevel(Level.ALL);
   }

   /**
    * Gives access to the logger for the component.
    * 
    * @return The logger for this component.
    */
   public final Logger getLogger() {
      return my_logger;
   }

   /**
    * Calls the listener and attempts to connect to a service.
    * 
    * @param a_serviceName
    *           the name of the service we wish to connect to.
    * @throws ServiceException
    *            if the named serice cannot be found. *
    * @throws IOException
    *            if the local proxy cannot be contacted.
    * @return a SocketWrapper round the desired service
    */
   protected final Socket getService(final String a_serviceName) throws IOException,
         ServiceException {
      Socket socket = null;
      // attempt to connect to the local bonjour proxy to get service descriptors
      // We want to get the service descriptors for an instance of construct
      // Iterate through them until we successfully resolve
      List serviceDescriptors;
      final String xmlServiceDescriptors = getXMLServiceDescriptors();
      if (xmlServiceDescriptors != null) {
         serviceDescriptors = getServiceDescriptors(xmlServiceDescriptors);

         // Sort the service descriptors to prefer the local host.
         Collections.sort(serviceDescriptors);

         // Now check for the named service in the set of descriptors
         socket = searchForAndConnectToService(serviceDescriptors, a_serviceName);
      }
      // At this point we either sucessfully connected to the service, or no service of that
      // name is known. If we were unsuccessful, we throw an exception
      if (socket == null) {
         throw new ServiceException("Found no running instances of a service with name: "
               + a_serviceName);
      }
      return socket;
   }

   /**
    * Iterate through a collection of services looking for a named service. We then try to
    * connect to that service.
    * 
    * @param the_services
    *           the collection of services to look through.
    * @param the_serviceName
    *           the name of the service we want to connect to.
    * @return a connected Socket to the service, or null if we could not obtain a connection.
    */
   private Socket searchForAndConnectToService(final Collection the_services,
         final String the_serviceName) {
      Socket commsSocket = null;
      boolean connected = false;
      final Iterator it = the_services.iterator();
      while (it.hasNext() && !connected) {
         final ServiceComponentDescriptor descriptor = (ServiceComponentDescriptor) it.next();
         if (descriptor.getName().equals(the_serviceName)) {
            try {
               commsSocket = new Socket(descriptor.getHost(), descriptor.getPort());
               connected = true;
            } catch (UnknownHostException uhe) {
               System.err.println("Unknown host: " + descriptor.getHost() + ":"
                     + descriptor.getPort() + uhe);
            } catch (IOException ioe) {
               System.err.println("Socket comms failure: " + descriptor.getHost() + ":"
                     + descriptor.getPort());
            }
            my_logger.info("Connected to "+the_serviceName+" on "+descriptor.getHost()+":"+descriptor.getPort());
         }
      }
      return commsSocket;
   }

   /**
    * Get the service descriptors in a Collection of ServiceComponentDescovery objects.
    * 
    * @param the_xmlDescriptors
    *           the service descriptors in XML format.
    * @return the collection of ServiceComponentDescriptor objects (which could be empty)
    */
   private List getServiceDescriptors(final String the_xmlDescriptors) {
      final List serviceDescriptors = new ArrayList();
      int startPos = the_xmlDescriptors.indexOf("<servicecomponentdescriptor>");
      while (startPos != -1) {
         final int endPos = the_xmlDescriptors.indexOf("</servicecomponentdescriptor>",
               startPos)
               + SERVICE_COMPONENT_DESCRIPTOR_END_TAG_LENGTH;
         final ServiceComponentDescriptor descriptor = ServiceComponentDescriptor
               .unmarshal(the_xmlDescriptors.substring(startPos, endPos));
         serviceDescriptors.add(descriptor);
         startPos = the_xmlDescriptors.indexOf("<servicecomponentdescriptor>", endPos);
      }
      return serviceDescriptors;
   }

   /**
    * Get the service descriptors in a big XML string.
    * 
    * @return the XML form of the service descriptors (which could just be
    *         &lt;services>&lt;/services> or null if we are not connected to an instance of
    *         Construct
    * @throws IOException
    *            if the local proxy cannot be contacted.
    */
   private String getXMLServiceDescriptors() throws IOException {
      Socket toConstruct = null;
      String result = null;
      final ExecutorService executorService = Executors.newSingleThreadExecutor();
      try {
         toConstruct = new Socket("localhost", LOCAL_CONSTRUCT_BONJOUR_PROXY_PORT);
         // Process message from discovery service
         final MessageReader messageReader = new MessageReader(toConstruct.getInputStream(),
                executorService);
         final Message message = messageReader.getMessage();
         result = message.getPayload();
      } catch (IOException ioe) {
         throw new IOException("Could not connect to the local proxy service. "
               + "Are you sure it is running?\nTry running \"telnet localhost 3826\" "
               + "from a command prompt to check.");
      } catch (InterruptedException e) {
         getLogger().warning(
               "InterruptedException " + "occured while " + "reading message: "
                     + e.getMessage());
      } catch (TimeoutException e) {
         getLogger().warning("TimeoutException while reading message: " + e.getMessage());
      } finally {
         if (toConstruct != null) {
            try {
               toConstruct.close();
            } catch (IOException e) {
               System.err.println("Could not close socket: " + e.getMessage());
            }
         }
         if (executorService != null) {
            executorService.shutdown();
         }
      }
      return result;
   }
}

File: COMP30050-2008/users/BHarty/construct/ruby/AbstractClientProxy.rb
# $Id: AbstractClientProxy.rb 3625 2007-03-15 12:36:26Z gstevenson $

# Copyright (c) 2006, Graeme Stevenson. All rights reserved.
# This file is part of Construct, a context-aware systems platform.
# Copyright (c) 2006, 2007 UCD Dublin. All rights reserved.
#
# Construct is free software; you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation; either version 2.1 of
# the License, or (at your option) any later version.
#
# Construct is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with Construct; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
# USA
#
# Further information about Construct can be obtained from
# http://www.construct-infrastructure.org

# Client-side proxy object that can talk to a Construct discovery service, find services and
# interact with them.
# 
# Author: Graeme Stevenson (graeme.stevenson@ucd.ie)
#
require 'socket'
require 'SocketWrapper'
require 'Message'
include Socket::Constants

class Service

attr_reader :name, :host, :port
  
  def initialize(name, host, port)
   @name = name
   @host = host
   @port = port
  end
  
  def to_s()
    return @name.to_s + "-" + @host.to_s + ":" + @port.to_s
  end
end

class AbstractClientProxy

  # The port on which the Construct bonjour proxy is running.
  LOCAL_CONSTRUCT_BONJOUR_PROXY_PORT = 3826
  
  #The length of the </serviceComponentDescriptor> tag.
  SERVICE_COMPONENT_DESCRIPTOR_END_TAG_LENGTH = 28

   # The identifier representing an OK status message.
   STATUS_OK = "600"
   
    # The identifier representing an ERROR status message.
    STATUS_ERROR = "610";

   # The identifier representing an UNKNOWN status message.
  STATUS_UNKNOWN = "650";

private

  # Get the service descriptors in a big XML string.
  # Return the XML form of the service descriptors (which could just be
  # &lt;services&gt;&lt;/services&gt; or nil if we are not connected to an instance of
  # Construct
  def getXMLServiceDescriptors() 
    begin
      #socket = TCPsocket.open("localhost", 3826) # open socket
      socket = Socket.new( AF_INET, SOCK_STREAM, 0 )
      sockaddr = Socket.pack_sockaddr_in( 3826, 'localhost' )
      socket.connect( sockaddr )
      descriptor = Message.new(socket).payload()
    rescue
      descriptor = nil # an error occured
    ensure
	socket.close if socket # close socket
    end
  end

    # Get the service descriptors in an Array of ServiceComponentDescovery objects.
    def getServiceDescriptors(the_xmlDescriptors)
      serviceObjects = Array.new
      serviceDescriptors = Array.new
      startPosition = the_xmlDescriptors.index('<servicecomponentdescriptor>')
      while(startPosition) do # Add the XML for each service into the array
         endPosition = the_xmlDescriptors.index('</servicecomponentdescriptor>', startPosition) + SERVICE_COMPONENT_DESCRIPTOR_END_TAG_LENGTH
         serviceDescriptors << the_xmlDescriptors.slice!(startPosition..endPosition) #Slice up the string and add results to the array
         startPosition = the_xmlDescriptors.index('<servicecomponentdescriptor>')
      end

      # Extract the service name, host name, and port
      serviceDescriptors.each{|desc| 
        name = desc.match('<name>[\s\w]+<\/name>').to_s if serviceDescriptors[0]
        host = desc.match('<host>[\s\w.]+<\/host>').to_s if serviceDescriptors[0]
        port = desc.match('<port>[\s\w]+<\/port>').to_s if serviceDescriptors[0]
        name = name.slice(6, name.length() -13) if name
        host = host.slice(6, host.length() -13) if host
        port = port.slice(6, port.length() -13) if port
        # Create a service if all three properties are present
        serviceObjects << Service.new(name, host, port) unless ((name == nil) or (host == nil) or (port == nil))
       }
    return serviceObjects  
  end

  # Iterate through a collection of services looking for a named service.
  # We then try to connect to that service.
  def searchForAndConnectToService(the_services, the_service_name)
    commsSocket = nil
    connected = false
    candidates = the_services.find_all { |service| service.name == the_service_name}
    
    # try to connect to a service until we are sucessful
    candidates.each{|candidate|
    begin
      commsSocket = Socket.new( AF_INET, SOCK_STREAM, 0 )
      sockaddr = Socket.pack_sockaddr_in( candidate.port, candidate.host)
      commsSocket.connect( sockaddr )
      connected = true
    rescue
       puts "Failed to connect: " + candidate.name + " " + candidate.port
    end   
     break if connected
    }
    return commsSocket
  end	

public
  # Obtains a connection to a service and places it in a wrapper.
  def getService(a_service_name)
    wrapper = nil
    desc = getXMLServiceDescriptors()
    if desc
      services = getServiceDescriptors(desc) 
      socket = searchForAndConnectToService(services, a_service_name)
      if socket
        begin
          wrapper = SocketWrapper.new(socket)
        rescue
          puts "Error when creating socket wrapper"
        end
      end    
    else
      raise "Local construct bonjour proxy not running on port 3826"
    end
    return wrapper
 end

end


